Font Library

matiasbenedetto opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->
:warning: this is a work-in-progress draft.

## What?
This PR implements font management capabilities into the Gutenberg editor.


#### Screencast: Activate, de-activate and install fonts:
https://github.com/WordPress/gutenberg/assets/1310626/e5caa889-dd43-4b60-8dcd-d58e52d40859

#### Screencast: Persisting fonts installed across themes:

https://github.com/WordPress/gutenberg/assets/1310626/ff447a5c-eedc-4de6-aa47-11a16a79beb2

#### Screencast: Uploading local font assets:



https://github.com/WordPress/gutenberg/assets/1310626/47610796-22c2-407e-bf7b-b1a1ba26e8f6





## Why?
Refer to https://github.com/WordPress/gutenberg/issues/45271

## How?
By the implementation of a Font Library.

### What is the font library?
The Font Library consists of a font manager for WordPress. The Font Library is available globally, independently of the theme activated, similar to the Media Library. This list of installed fonts and their assets are site-wide available and the users can select the fonts activated (available in the editor) for each theme.

Using the Font Library the users can:
- Install fonts
    - Install Google fonts
    - Install local fonts
- Uninstall fonts
- Activate/deactivate fonts

In this PR we implement the font installation from Google fonts and local fonts, but I think this implementation will make it easy to create third-party plugins that offer a list of fonts from which users can pick and install the ones they want to use.

### How it works:
- **Install fonts:** add the font family definition to the installed fonts. list and download the font face assets to `/wp-content/fonts`).
    - **Install Google fonts:** add the font families definition to the list and downloads the Google fonts assets to `/wp-content/fonts`. The assets from Google are called just at the time of previewing fonts in the font library. After the fonts is installed no other request to Google or other third-party servers will be made because the font assets are hosted locally in the `/wp-content/fonts` folder.
    - **Install local fonts**  The font assets are uploaded from the users devices to the `wp-content/fonts` folder and the font definitions are added to the list of installed fonts.
- **Activate/deactivate fonts:** add/remove font-family definitions from the font families list defined in Global Styles/Settings. The active fonts are the fonts that will be available to use in the editor. As we are using the Global Styles data the users are able to activate/deactivate fonts for each theme. 
- **Uninstall fonts:** remove the font family definition from the list of installed fonts and delete the assets from `/wp-content/fonts` folder. Only the non-active fonts from the user-installed fonts can be uninstalled. Theme fonts will be able to be deactivated but not uninstalled.

## Known issues:
- **Dynamically load font in the editor iframe**:  When users activate a recently installed font in the editor the font is not loaded on the iframe editor because it does not reload on global styles font families change. We need to fix this to preview the font on the editor just right after the font is activated without the need of reloading the page.
- **Removing font activated in another theme:** If users remove and user-added font not active on the current theme but active in another theme the assets will be removed but the font definitions will persist. We should define the expected behavior around this.
- **More visual hints**: we would need to add more visual hints for example to notify a font was successfully installed.

### How the data is persisted?

1. **User-installed fonts:** the list of installed fonts is saved in a custom post type, `wp_fonts_library`, in the database, similar to how Global Styles. That is an addition to this PR. It is motivated by the necessity of having a list of fonts that is independent of the theme activated and currently, we don't have that ability. 

2. **Theme installed fonts:**  the list of theme fonts comes from the theme origin of the Global Styles/Settings.

3. **Active fonts list:** From the previous 2 sources the user can add font definitions to a third list persisted in the Global Styles custom origin. This is the list of active fonts for each theme. If this list is not defined in Global Styles, all the theme fonts are active.

## How to test
Use the feature as features in the screencast.
**All the feedback about the design, behavior, and data-persistence of the feature is welcome!** 

## About the code implementation
I consider that is not ready and it doesn't worth making a code review yet because the implementation is sketched to test if this work is on a good path. Currently, we are implementing just the happy path, there is not much error checking, no security checks, not performance optimization and no tests.

## Previous explorations around this:
https://github.com/WordPress/gutenberg/pull/46332
https://github.com/WordPress/gutenberg/pull/50102

matiasbenedetto committed:
adding a mockup of the font library

matiasbenedetto committed:
adding endpoint for google fonts api font list

matiasbenedetto committed:
more UI work for fonts library

matiasbenedetto committed:
added computed categories to google fonts api data

matiasbenedetto committed:
UI work

matiasbenedetto committed:
installing complete families of google fonts

matiasbenedetto committed:
activate and deactivate font families and font variants

matiasbenedetto committed:
adding global styles save button

matiasbenedetto committed:
transforming google fonts api response in to themejson fontFamilies format

matiasbenedetto committed:
More UI Work

matiasbenedetto committed:
More UI Work

matiasbenedetto committed:
implementing toggleInstallFont and installedFontsOutline

matiasbenedetto committed:
UI implementation work

matiasbenedetto committed:
google fonts search

matiasbenedetto committed:
font variant styles

matiasbenedetto committed:
select control for categories

matiasbenedetto committed:
discard font library changes when switching tabs in the font library modal

matiasbenedetto committed:
activated fonts

matiasbenedetto committed:
load theme fonts assets

matiasbenedetto committed:
UI work for font previews

matiasbenedetto committed:
fixing logic  for select/unselect variants to install

matiasbenedetto committed:
UI work

matiasbenedetto committed:
saving google fonts working

matiasbenedetto committed:
handling font uninstall

matiasbenedetto committed:
toggle activation for system fonts (fonts without font faces)

matiasbenedetto committed:
adding preview controls to installed fonts tab

matiasbenedetto committed:
fixing toogle activate fonts and font faces

matiasbenedetto committed:
fix available fonts logic

matiasbenedetto committed:
UI work

matiasbenedetto committed:
avoid the uninstall button if the font was added by a theme

matiasbenedetto committed:
adding a flag to the theme installed fonts to avoid being removed

matiasbenedetto committed:
fixing is selected logic for google fonts

matiasbenedetto committed:
uninstall fonts api endpoint

matiasbenedetto committed:
dealing with fonts without name

matiasbenedetto committed:
avoid mixing custom and theme fonts

matiasbenedetto committed:
avoid mixing custom and theme fonts

matiasbenedetto committed:
google fonts api downloader and updater

matiasbenedetto committed:
fixinf card styles with DEBUG false

matiasbenedetto committed:
removing unused function

matiasbenedetto committed:
remove the font families with no font faces

matiasbenedetto committed:
removing unused utils

matiasbenedetto committed:
removing unused imports

matiasbenedetto committed:
using slug instead of name to find fonts

matiasbenedetto committed:
simplify function

matiasbenedetto committed:
moving fonts library backend files to its own folder

matiasbenedetto committed:
adding phpdocs

matiasbenedetto committed:
JS lint

jasmussen commented:
Fantastic work, nice to see something functional. Here's a GIF:

![fonts](https://github.com/WordPress/gutenberg/assets/1204802/0900cbe1-28ce-4cd8-a5dd-5a7bb5248f02)

It diverges quite a bit from the mockups initially provided. Was the modal easier to implement than the drilldowns? 

At this point I'm happy to land something like this in the next 5 days so we have something for 6.3, modal or otherwise, and we can work with this. But there are some important details from the mockups that we should come back to. The font selector is one example that can easily be separated out. We can also separately look at font sets. But a lot of details around adding font weights, variable width font weights, and the ability to supply on context of each font, a warning when picking too many font weights. 

The drilldown also provides something for each font in the ItemGroup to link to. The fact that they are inert here is a little bit confusing:

<img width="277" alt="Screenshot 2023-05-25 at 10 46 49" src="https://github.com/WordPress/gutenberg/assets/1204802/e4d6a99f-5545-43e9-9eb2-fd1de981d746">

The ellipsis has 3 options but all lead just to tabs inside the modal. If we keep this modal approach for 6.3, we remove the dropdown entirely, and perhaps with the "settings" icon, simply open the modal on the "manage" tab.

We should also probably avoid the term "Fonts Library", as we are eventually getting a Library item in the site view: #50397. Perhaps we can just call it "Fonts" in the interim.

I'm not sure it's valuable to use elevation to distinguish between selected and unselected fonts:

<img width="1086" alt="Screenshot 2023-05-25 at 10 50 48" src="https://github.com/WordPress/gutenberg/assets/1204802/cb07366e-d302-4dfc-be5d-8f63ec61537d">

It's also not entirely intuitive that you have to click the font card itself, not just the checkmark, to get access to font weight management:

<img width="581" alt="Screenshot 2023-05-25 at 10 53 44" src="https://github.com/WordPress/gutenberg/assets/1204802/36a425c5-e400-4c99-86cc-75c9b5018959">

Let me know if this feedback is useful. But I'd encourage us finding a simple path to the core functionality of adding and removing font weights, locally and from Google, much moreso than the preview features, so hopefully we can land something very soon! Nice work.

matiasbenedetto commented:
Thanks for the feedback @jasmussen !
A few comments/questions about it:

> It diverges quite a bit from the mockups initially provided. Was the modal easier to implement than the drilldowns?

We explored the modal in mockups [before](https://github.com/WordPress/gutenberg/pull/46332#issuecomment-1367245772) and I think it makes much more sense than the sidebar, mainly because in the sidebar there is not enough space to preview the fonts properly. In terms of code complexity, there is not much difference but I think the UX is better.


> But a lot of details around adding font weights, variable width font weights, and the ability to supply on context of each font, a warning when picking too many font weights.

Could you add more details about this, please?

> adding font weights.

Not sure what you mean. Currently this is possible, if you click a font family card in the Google fonts tab you can pick the desired font weights you want to install. Screenshot:

![image](https://github.com/WordPress/gutenberg/assets/1310626/893474df-8581-41da-968c-2ad7ee8321b1)


> variable width font weights

That will be possible using the local fonts upload because variable fonts are just any other font file. But the installing variable fonts from Google Fonts won't be possible yet because variable fonts are not available yet on Google Fonts API. The API delivers them as smaller regular font assets for each variant instead of a larger file containing the variable version.

> supply on context of each font

I'm not sure what you mean.

> A warning when picking too many font weights.

Picking too many font weights refers to installing too many fonts or activating too many fonts?



> The drilldown also provides something for each font in the ItemGroup to link to.

What do you think of opening the modal showing the font family details of the font clicked ? I mean this screen:

![image](https://github.com/WordPress/gutenberg/assets/1310626/3ce940b4-7931-4aba-b065-1241a0f708d7)


> We should also probably avoid the term "Fonts Library", as we are eventually getting a Library item in the site view: https://github.com/WordPress/gutenberg/issues/50397. Perhaps we can just call it "Fonts" in the interim.

Good, I will replace the user facing name. If we get a definitve name I can rename the source code too.

> I'm not sure it's valuable to use elevation to distinguish between selected and unselected fonts

Ok, I was not sure about that either. I will remove it.

> It's also not entirely intuitive that you have to click the font card itself, not just the checkmark, to get access to font weight management

The checkmark is used to enable/disable the entire font family. I think it's important to have a way to enable/disable the entire font family. I think the use case of disabling/enabling entire font families will be more used than enabling or disabling certain font variants in particular. How can we improve that?


> The core functionality of adding and removing font weights, locally and from Google.

We have the font variants addition form Google currently working. I will add the local upload asap.


Sorry for the numerous questions and thanks again for the useful feedback!

matiasbenedetto committed:
removing card elevation

jameskoster commented:
A couple of high-level feedback items:

* I quite like the modal experience. It's a good way to preview many fonts/weights at once, and can be invoked from both the styles panel in edit view, and the new Library panel in site view which seems handy.
* Agree with Joen that the itemGroup is a little strange with no interaction. Do we need it? A simple "Manage fonts" button somewhere could be adequate – font management doesn't feel like a very common task. 
* There's something about managing installed fonts, and installing new ones at the same time that feels a bit off currently. It's easy to lose track of what the Save action will include. I would consider:
  * Separate modals for adding / managing, or
  * Instant installation. So if I toggle a font on in the "Google fonts" tab, it immediately appears in the "Installed fonts" tab (perhaps with a `Spinner` while it loads). 
 * After installing a Google font I still see it listed in the "Google fonts" tab, and I'm able to repeat the installation process. Installed Google fonts should either be hidden, or marked as installed. 
 * Checkboxes on the "Installed fonts" tab need an indeterminate state for partial variant selection.

Returning to the aforementioned Library, here's how you might add / manage fonts there:

<img width="1521" alt="Library" src="https://github.com/WordPress/gutenberg/assets/846565/1d0dd06a-3da3-40b5-b1a1-fd32ce068bb9">

jasmussen commented:
> Could you add more details about this, please?
> The checkmark is used to enable/disable the entire font family. I think it's important to have a way to enable/disable the entire font family. I think the use case of disabling/enabling entire font families will be more used than enabling or disabling certain font variants in particular. How can we improve that?

That was one of the flows that was honed a fair bit in #45271, which is why it would've been nice to converse a little during implementation to make sure it wasn't lost. Specifically, the combination of using the `Navigator` combination with the `ItemGroup` would let you see all the actual font files that were installed, in terms of every weight:

<img src="https://user-images.githubusercontent.com/1204802/233346829-a05efab0-5c08-483c-96d4-ae6e8b07f644.png">

There would still be a checkbox to select all if that's really what you want. But it would be good to nudge our users towards being responsible with font usage, and not install too many weights at once, bandwidth is still an issue in large parts of the world. The drilldown for each, and requireing you to pick weights (or all) gives an opportunity to explain that.

Similarly, the first ItemGroup was meant to show installed fonts, to Jay's point if we go with the modal it's not going to be that useful and we'll need a different singular button for managing fonts. 

The upload flow could also still be relevant, despite the modal:

<img src="https://user-images.githubusercontent.com/1204802/237756501-ad34bdd5-d222-4812-b8d8-324aa82bf93d.png">

In order to land something for 6.3, I think Jay's approach above makes sense.

matiasbenedetto committed:
adding lib-font library to read font metadata tables

matiasbenedetto committed:
implementing local fonts upload

matiasbenedetto committed:
CSS fixes for font previews

matiasbenedetto commented:
:wave: I added the upload local fonts implementation, and I'm working on addressing your feedback about splitting the modal contents In 2.


https://github.com/WordPress/gutenberg/assets/1310626/47610796-22c2-407e-bf7b-b1a1ba26e8f6

matiasbenedetto committed:
add partial selection state for font cards

matiasbenedetto committed:
Merge branch 'trunk' into try/font-library

matiasbenedetto committed:
Updating endpoints permissions checks

jasmussen commented:
Here are preliminary mockups that leverage a modal. I've omitted most font previewing functionality from this iteration, to start with less, get that right, and then add as it becomes clear it's necessary. I'm also not completely feeling these mockups, so I would appreciate a sanity check from @jameskoster or @richtabor or @SaxonF, so I can refine them more, but as far as keeping the modal and the tabs, this is the direction I'm thinking. It includes:

* tab bar and back button system
* interface for deleting or activating or deactivating
* a basic list of installed fonts in the GS inspector, linked to the modal
* a separate "installed fonts" tab

Add from Google:

<img width="4800" alt="Add fonts from Google" src="https://github.com/WordPress/gutenberg/assets/1204802/9519592c-adc3-4fc6-bcbc-a6cf403256e8">

Manage fonts:

<img width="4240" alt="Manage fonts" src="https://github.com/WordPress/gutenberg/assets/1204802/29d9c9d6-99b5-47cb-9c55-6bc955e545a5">

Deleting or deactivating:

<img width="5760" alt="Deleting or deactivating fonts" src="https://github.com/WordPress/gutenberg/assets/1204802/45e1590e-6933-48b1-9b3b-6339ef3a5ba7">

Uploading fonts:

<img width="7840" alt="Upload fonts" src="https://github.com/WordPress/gutenberg/assets/1204802/18281c97-83d7-4da8-8002-efecc02c4e9d">

matiasbenedetto committed:
Adding some error handling on api endpoints

matiasbenedetto committed:
Adding error handling and comments in Fonts library php class methods

jasmussen commented:
Thanks for the ongoing work. To keep us nimble, I've gone ahead and updated the original issue with the aforementioned mockups. I would also, for the sake of landing a first version of this, suggest that we can stay nimble and not even reach the full picture for V1:

* Google Fonts, Upload, and Installed tabs in the modal
* Installed tab only shows up when you have installed fonts, and this is where you manage activation and deactivation and deletion.
* There’s no single “choose all font weights” checkbox, all font cards go directly to the individual weights list. That affords us to also show clearer activate/deactivate/delete buttons, as well as context on file sizes.
* I’ve omitted all previewing tools in the name of simplicity. It adds weight to the UI as is, and is more of an UX burden than it is a technical burden. That doesn’t mean we can’t add it back, but I would hide it in the name of simplicity for V1.

Let me know if that works!

matiasbenedetto committed:
Add mime type check for font file paths

matiasbenedetto committed:
Initialize Fonts Library api from it own filee

matiasbenedetto committed:
Adding prefix to installed font slugs to avoid slug collisions

matiasbenedetto committed:
Move fonts directory file permission check to permission callback of the updatable endpoints

matiasbenedetto committed:
Avoid duplicated files when selecting local fonts to upload

matiasbenedetto committed:
Replacing modal opening from dropdown to 2 buttons with icons

matiasbenedetto committed:
Fonts modal title change

matiasbenedetto committed:
changing main fonts list

matiasbenedetto commented:
Work in progress: I've been polishing the backend a little and did some UI work to replace the dropdown for the **+** and **Aa** buttons to open the different tabs:

https://github.com/WordPress/gutenberg/assets/1310626/4fb6f127-46e4-4be5-8bca-e178e9ac6c02

matiasbenedetto committed:
Removing action handler and google and installed font card components

matiasbenedetto committed:
Splitting the installed fonts in theme and user fonts sections

matiasbenedetto committed:
hiding preview options

matiasbenedetto commented:
More changes based on feedback:

- Splitting theme fonts and user-installed fonts in the installed fonts tab
- Removing preview options (custom demo text and font size of the previews)
- Display the number of variants active


![image](https://github.com/WordPress/gutenberg/assets/1310626/15e9daeb-a7e2-44e3-97ab-8ee91a1defe0)

jasmussen commented:
Outstanding work, thank you Matias. Here's a quick GIF that shows clicking to add a font, updating a font, and seeing the result reflected in the sidebar:

![fonts status](https://github.com/WordPress/gutenberg/assets/1204802/e3a3afd8-b435-46ab-abe6-74c0e4b36665)

This behavior is correct according to #45271, which means we are now ready to land a first version of this PR. For that to happen, we need a few things, beyond a rebase and marking the PR ready for review.

<img width="270" alt="Screenshot 2023-06-06 at 10 46 21" src="https://github.com/WordPress/gutenberg/assets/1204802/2096891f-3c7a-4123-8f5b-6185586e7d1a">

This list, it would be nice if clicking each of these fonts took you to the Installed Fonts > [Font name] drilldown. Let's also call the font files as "Styles" instead of "Variants". This isn't the strongest opinion in the world, we can revisit and go back to variants in the future if it's confusing, but I'd love to try this as a start. Not a blocker, but we should use the same term across so keep that in mind for feedback below.

Instead of "Saved", the button should say "Update", both in its disabled and enabled states, for any fonts you already have:
<img width="224" alt="Screenshot 2023-06-06 at 11 06 47" src="https://github.com/WordPress/gutenberg/assets/1204802/a2658277-67f4-407e-a429-20f0cf7dbfe1">

When adding fonts fresh, omit the download icon, and instead say "Add & Select":

<img width="283" alt="Screenshot 2023-06-06 at 11 08 19" src="https://github.com/WordPress/gutenberg/assets/1204802/da551a09-aa35-4241-910a-b04638a66d3b">

Let's add tooltips here:

<img width="87" alt="Screenshot 2023-06-06 at 10 46 24" src="https://github.com/WordPress/gutenberg/assets/1204802/417180ad-5e7e-4ea0-8c8f-a568c48c5187">

Can be:

* Plus — "Add fonts"
* Typography — "Manage fonts"

<img width="144" alt="Screenshot 2023-06-06 at 10 51 05" src="https://github.com/WordPress/gutenberg/assets/1204802/9a3fafa0-df7a-4c68-9f90-6a44a4b01675">

This tab can just be "Library".

On this one:

<img width="592" alt="Screenshot 2023-06-06 at 10 52 19" src="https://github.com/WordPress/gutenberg/assets/1204802/032f05c7-a82f-43c9-bf40-718a922e5e9a">

* You can remove the text above, "Fonts installed in your ..."
* User fonts should be above Theme fonts
* You can remove the "User fonts" subheading

On drilldowns:

<img width="785" alt="Screenshot 2023-06-06 at 10 53 22" src="https://github.com/WordPress/gutenberg/assets/1204802/2c5bfe0a-b0b9-41f9-b6fe-399d193f5f95">

You can replace the "IBM Plex Mono variants." text with a font-weights warning and a "STYLES" subheading, as shown here:

<img width="891" alt="Screenshot 2023-06-06 at 11 00 39" src="https://github.com/WordPress/gutenberg/assets/1204802/efe6cea5-3e97-463b-bab0-d7bfa58bc2f3">

On the Google Fonts tab, let's add a little introductory text, this:

<img width="583" alt="Screenshot 2023-06-06 at 10 57 49" src="https://github.com/WordPress/gutenberg/assets/1204802/a77a762c-699a-40a2-ab0e-c32556a6eb18">

With those changes in place, I think we are good to land **this PR as a first iteration**. 🎉

jasmussen commented:
Splitting into a separate comment, here are changes we can do as followups. **None of the chagnes below should block this PR from landing**. 

We should remove the "Upload" tab, and integrated it into the Library page:

<img width="105" alt="Screenshot 2023-06-06 at 10 46 58" src="https://github.com/WordPress/gutenberg/assets/1204802/7f8e163f-2d4a-4674-94d7-ab8b7fdb1d67">

<img width="933" alt="Screenshot 2023-06-06 at 10 55 27" src="https://github.com/WordPress/gutenberg/assets/1204802/d0f31156-d491-4d74-aac7-d2b054fd0de3">

The entire library page can be one big dropzone for uploading fonts, but we can still have a button below.

The search and category components can be updated to match this design:
<img width="786" alt="Screenshot 2023-06-06 at 10 58 23" src="https://github.com/WordPress/gutenberg/assets/1204802/4759d7a3-5cb1-47f7-9278-69ba8dbf00ab">

The list view can be updated to be vertical, like this:

<img width="800" alt="Screenshot 2023-06-06 at 10 58 51" src="https://github.com/WordPress/gutenberg/assets/1204802/7267e492-702b-4228-bf6b-c7e18c7c3322">

The modal is huge:

<img width="2032" alt="Screenshot 2023-06-06 at 11 02 42" src="https://github.com/WordPress/gutenberg/assets/1204802/853a8904-ac83-4c57-91f6-9c595b262cbd">

It would be good to update the metrics so there's 64px margin all around the modal, and a max-width of 750 (or the nearest nice number):

<img width="1440" alt="Overview" src="https://github.com/WordPress/gutenberg/assets/1204802/aae083fc-edcc-4667-9453-b4effccfcd08">

The drilldown list view can ItemGroup-like borders, same as the list view shown just now, but with checkboxes to choose weights, including a checkbox to "select all".

We can show the number of weights in the modal footer, and include a delete button:

<img width="892" alt="Screenshot 2023-06-06 at 11 04 32" src="https://github.com/WordPress/gutenberg/assets/1204802/3f2251f0-8dea-4d67-96c8-d6ee76475530">

The delete button would just uncheck all, and "Update" at the same time.

When you're adding a font fresh, we can still show the number of styles selected, and omit the delete button of course. The button should say "Add & Select":
<img width="925" alt="Screenshot 2023-06-06 at 11 06 15" src="https://github.com/WordPress/gutenberg/assets/1204802/6d64039f-6795-4931-bc3d-f609a80c08e6">

matiasbenedetto committed:
Moving user fonts at the top of library modal and removing sub heading

matiasbenedetto committed:
Adding tooltip to panel buttons

matiasbenedetto committed:
changing installed fonts tab title to library

matiasbenedetto committed:
moving modal is open state and installed font selected state to the context provider to allow open installed font details (variants) from the sidebar

matiasbenedetto committed:
Update save buttons text

matiasbenedetto commented:
> Let's also call the font files as "Styles" instead of "Variants"

I think there is something that may be problematic around the wording. The mockups use "Styles" and "Weights" to refer to the members of a font family. I think that is not the best word because a font style designates if the font is italic or normal, and weight designates how bold a font face is. I think "Variant" is a better term for this because it is agnostic about these font characteristics. Any other term that doesn't refer to a specific font characteristic could work too.

matiasbenedetto committed:
Update tab texts

matiasbenedetto committed:
Updating tab layout title

matiasbenedetto commented:
Thanks @jasmussen for the feedback and the mockups updates! 
I have only one comment about the working of variants posted before. Apart from that in the latests commits I implemented the items you added in [this comment.](https://github.com/WordPress/gutenberg/pull/50927#issuecomment-1578255822).

I think we would need 3 things to mark the PR as ready for review:
1. Fix a bug around the drill-down when opened from the sidebar.
2. Removed non-allowed properties in the JSON received by the endpoints.
3. Define when a font can be uninstalled (remove the font family definition and assets).

1 and 2 should be easy to fix, I will implement that tomorrow.

3 is more tricky because we need to define what happens when a user tries to remove a font from the library activated in another theme. For example, the user installs and activates the font "Inter" using theme A. After that, the user switches to theme B and deletes "Inter". If the user activates theme A again the font definition will be there but the font assets won't. As a result, the text elements with "Inter" applied won't render Inter but fallback typography.

We could search the global styles for all the themes to know if the font is activated or not before allowing it to delete the font but it seems like an expensive task if the user has a lot of themes installed. What are your thoughts on this?

matiasbenedetto commented:
We could discuss this later, but the changes for the next iteration proposed in [this comment](https://github.com/WordPress/gutenberg/pull/50927#issuecomment-1578256983) seem to let out all the preview options that we remove from this iteration to add later. Apart from that, reducing the width of the modal has no practical advantages because the rest of the screen is still not usable. I received feedback from designers that the preview options are a useful tool for picking font families and my experience also a user tells me that the font previews are a key part of choosing fonts for a project so I would consider adding them. Without this tool the user should go back and forth out of the editor to be able to play with the typographies and return to the editor with a decision made. I guess that is not ideal, more when we can implement the feature easily and we received good feedback about this aspect of the Create Block Theme interface.

matiasbenedetto committed:
Changing upload local fonts button

matiasbenedetto committed:
Select the installed font insted of the font stored in custom global styles when a font from the sidebar is selected.

We need to this because otherwise the font from custom fonts is used and when a font variant is unselected it is deleted when it is unselected and the user saves the modal.Doing this we can still have record of all the installed variants that can be activated or de-activated.

matiasbenedetto committed:
moving where alphabetical orderingis done

jasmussen commented:
> I think "Variant" is a better term for this because it is agnostic about these font characteristics. Any other term that doesn't refer to a specific font characteristic could work too.

Sounds fine, we can go with variant, we can always revisit. 

> Fix a bug around the drill-down when opened from the sidebar.
> Removed non-allowed properties in the JSON received by the endpoints.
> Define when a font can be uninstalled (remove the font family definition and assets).

Is there any chance we can split 3 out into a separate PR so we can land this? Perhaps you can expand on the problem and help me understand better.

> We could discuss this later, but the changes for the next iteration proposed in https://github.com/WordPress/gutenberg/pull/50927#issuecomment-1578256983 seem to let out all the preview options that we remove from this iteration to add later. 

These were intentionally omitted simply for speed of iteration. I think there are opportunities to revisit these in the future, but we have to implement them in a way that gels with the existing block editor design. I fully expect we can add these things back, but we have to do it using existing design patterns, so I'll help follow up with that. So the most important part isn't to remove these, but to find a configuration of this PR we can land today or tomorrow, and split the rest into followups.

> Apart from that, reducing the width of the modal has no practical advantages because the rest of the screen is still not usable.

The subpage of "Installed fonts" due to its subheadings works best as a vertical list, without the cards layout. For that, the smaller modal width benefits legibility. 

Just as with the previewing tools, the important bit here isn't to remove features, but to land a minimal and functional first version in the next day or two, and it does not preclude adding this back in followups, and I'll be happy to help.

jasmussen reviewed (approved):
Taking it for a quick review, here's a GIF of the flow of managing fonts, and uninstalling one:

![font manager](https://github.com/WordPress/gutenberg/assets/1204802/52b8368e-a158-4c79-a25b-00c402209914)

One bug I encountered with a theme font is that there's a missing title, not sure what's up there:

<img width="372" alt="Screenshot 2023-06-07 at 10 57 02" src="https://github.com/WordPress/gutenberg/assets/1204802/9674cc62-d10e-49a7-9a81-23bffd4fac0b">

<img width="845" alt="Screenshot 2023-06-07 at 10 57 28" src="https://github.com/WordPress/gutenberg/assets/1204802/6bc4a2f4-bea5-4bac-a65d-6c87c19c7636">

We should follow up with integrating the upload tab into the library, rather than having it be separately, but that can be a followup.

But I think this is good for a final review and looks good for a first version on my end.

matiasbenedetto committed:
Using custom and theme fonts from context to make profit of alphabetical sorting made there

matiasbenedetto committed:
Set fallback values for font familiies definitions

matiasbenedetto committed:
encode font asset url if it is not url encoded

matiasbenedetto committed:
Adding infinite scroll to fonts grid component

matiasbenedetto committed:
Adding font families json sanitization using WP_Theme_JSON capabilities

matiasbenedetto commented:
During the last WordCamp Europe I met @youknowriad and @oandregal and we made some real-life code review of this PR. They kindly provided this feedback that I will address in the next few days. Thanks, guys!

The feedback was mainly about leveraging certain abstractions that WordPress has in place with the objective of avoiding a custom implementation. These are my notes about the input they gave me:

### Frontend:
- [x] **Refactor the use of useSetting hook to get the current active fonts.**  ( Implemented in https://github.com/WordPress/gutenberg/pull/50927/commits/bd3ab6d41b3830ff9fd5a3b9fe20a6dc73d3ecd5 )
- [x] **Move lib-font dependency to package specific dependencies.**
- [x] **Try to find another library or a way to avoid changing Webpack config using lib-font.** 
Lib font with minor changes was added to the bundle to avoid changing the webpack config ( https://github.com/WordPress/gutenberg/pull/50927/commits/12b4ac154701479221945d6c197940071011fd0b ) 
- [ ] **Explore getEntityRecord to avoid a custom resolver.**

### Backend:
- [ ] **Find a way to gitignore google-fonts json file.**
- [ ] **Split WP_Fonts_library class between rest endpoints (new class extending Rest class) and the rest of the code.**
- [ ] **Explore the use of Register post type to create the fonts library post instead of creating the post manually.**
- [X] ~**Consider using attachments functionality. Are the fonts a new type of attachment?**~
This was considered and discarded. Quoting @hellofromtonya :

> IMO an "attachment" doesn't make sense for fonts. Why?
Each font is not attached to a "post" but rather entities within a post, such as the template or a block or even globally.
[Unlike images](https://developer.wordpress.org/themes/template-files-section/attachment-template-files/#attachment-php), I don't see a use-case for rendering a single font attachment.php webpage on the front-end.

- [ ] **Split the font library database persistence. Possibly use one post for each font family.**

spacedmonkey commented on lib/experimental/fonts-library/class-wp-fonts-library.php:
```diff
@@ -0,0 +1,618 @@
+class WP_Fonts_Library {
+    public function __construct() {
+        $this->wp_fonts_dir = path_join( WP_CONTENT_DIR, 'fonts' );
+        $this->relative_fonts_path = site_url('/wp-content/fonts/', 'relative');
```
```suggestion
        $this->relative_fonts_path = content_url('/fonts/');
```

oandregal commented:
@matiasbenedetto just one thing that may be already captured by your notes. The theme fonts and the font library fonts should live each in their own place. For example, as it stands today, using TwentyTwentyThree and adding a font from the google library, this is how the data is stored (see theme.json in database or, in the editor, look at the `core/block-editor` store, under `settings.__experimentalFeatures.typography.fontFamily`):

![image](https://github.com/WordPress/gutenberg/assets/583546/78904d69-76df-484f-8eea-1f2064b9acb8)

Note how the fonts defined by the theme are present in both `theme` and `custom` while they should only be present in `theme`.

matiasbenedetto committed:
remove lib-font npm dependency and include it in the source code with small modifications to make it work in the browser

matiasbenedetto commented:
> @matiasbenedetto just one thing that may be already captured by your notes. The theme fonts and the font library fonts should live each in their own place. For example, as it stands today, using TwentyTwentyThree and adding a font from the google library, this is how the data is stored (see theme.json in database or, in the editor, look at the `core/block-editor` store, under `settings.__experimentalFeatures.typography.fontFamily`):


Yes! Thank you @oandregal for pointing this out again. That change is reflected in the review notes. That mistake was motivated by believing the behavior described in this bug report ( #51487 ) was the correct behavior. Initially, I thought that the custom font families were overriding the theme ones as it seems to happen now in the font picker, but that is a bug of the picker and not how the theme.json data works.

matiasbenedetto committed:
use content_url instead of site_url

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

matiasbenedetto commented:
In [this commit ](https://github.com/WordPress/gutenberg/pull/50927/commits/12b4ac154701479221945d6c197940071011fd0b) I removed the `lib-font` [library](https://github.com/Pomax/lib-font) (used to read the local uploaded fonts metadata) dependency and added the source code that we use of the library to the Gutenberg codebase with a few changes that make it work in the browser with no errors and without changing the Webpack config as @youknowriad asked. Probably including the code in the GB codebase is not the best way to do this but it showcases how can we make it work. Input about where we could put this code is welcome!

spacedmonkey commented on lib/experimental/fonts-library/class-wp-fonts-library.php:
```diff
@@ -0,0 +1,618 @@
+ */
+class WP_Fonts_Library {
+    public function __construct() {
+        $this->wp_fonts_dir = path_join( WP_CONTENT_DIR, 'fonts' );
+        $this->relative_fonts_path = content_url('/fonts/');
+
+		$this->rest_base        = 'fonts_library';
+		$this->namespace        = 'wp/v2';
+
+        $this->post_type        = 'wp_fonts_library';
+        $this->post_name        = 'wp_fonts_library';
+        $this->post_title       = 'Fonts Library'; // Do not translate
+        $this->post_status      = 'publish';
+
+        $this->base_post_query  = array  (
+            'post_type'      => $this->post_type,
+            'post_title'     => $this->post_title,
+            'post_status'    => $this->post_status,
+            'post_name'      => $this->post_name,
+            'name'           => $this->post_name,
+            'posts_per_page' => 1,
+        );
+	}
```
This property should all be defined. PHP 8.1+ does not support dynamic properties.

matiasbenedetto replied:
added here: https://github.com/WordPress/gutenberg/pull/50927/commits/95ced36dbc6185b31435371ff9073a5ae60f8268

spacedmonkey commented on lib/experimental/fonts-library/class-wp-fonts-library.php:
```diff
@@ -0,0 +1,618 @@
+ * @since x.x.x
+ * 
+ */
+class WP_Fonts_Library {
```
```suggestion
class WP_Fonts_Library extends WP_REST_Controller {
```

spacedmonkey commented on lib/experimental/fonts-library/class-wp-fonts-library.php:
```diff
@@ -0,0 +1,618 @@
+            'post_status'    => $this->post_status,
+            'post_name'      => $this->post_name,
+            'name'           => $this->post_name,
+            'posts_per_page' => 1,
```
Do we need to prime post, term meta caches? Remove found rows.

matiasbenedetto committed:
list all the font families from theme json

matiasbenedetto committed:
Extend WP_Rest_Controller

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

anton-vlasenko commented on lib/experimental/fonts-library/class-wp-fonts-library.php:
```diff
@@ -0,0 +1,618 @@
+            return false;
+        }
+
+        return wp_delete_file( $file_path );
```
The `wp_delete_file()` function doesn't return anything, so this class method doesn't seem to work as intended.

matiasbenedetto replied:
Good catch. refactored this function in this commit: https://github.com/WordPress/gutenberg/pull/50927/commits/645626f1f8b9206c268a4dbc7f3b05ea25e17ac6

anton-vlasenko commented on lib/experimental/fonts-library/class-wp-fonts-library.php:
```diff
@@ -0,0 +1,618 @@
+		'woff2' => 'font/woff2',
+	);
+
+    function register_routes () {
```
```suggestion
    public function register_routes () {
```
Let's explicitly declare the visibility for methods in this class.
[Visibility should always be declared](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#visibility-should-always-be-declared).

As I understand it, the methods in this class should be either public or protected unless you want to forbid extending your methods in the descendant classes. Then they should be specified as `private`.
Since `register_routes()` is inherited from the parent class, it should have the same visibility, i.e., public.

anton-vlasenko commented on lib/experimental/fonts-library/class-wp-fonts-library.php:
```diff
@@ -0,0 +1,618 @@
+    function delete_font_face_assets ( $font_face ) {
+        if ( is_array( $font_face['src'] ) ) {
+            foreach ( $font_face['src'] as $src ) {
+                $was_assets_removed = $this->delete_asset($src);
```
This is not going to work as expected since `WP_Fonts_Library::delete_asset()` always returns null.

matiasbenedetto replied:
Good catch. I refactored this here: https://github.com/WordPress/gutenberg/pull/50927/commits/645626f1f8b9206c268a4dbc7f3b05ea25e17ac6

anton-vlasenko commented on lib/experimental/fonts-library/class-wp-fonts-library.php:
```diff
@@ -0,0 +1,618 @@
+            if ( isset( $font['fontFace'] ) ) {
+                $new_font_faces = array();
+                foreach ( $font['fontFace'] as $font_face ) {
+                    $new_font_face;
```
This line of code doesn't seem to be needed.
```suggestion
```

anton-vlasenko commented on lib/experimental/fonts-library/class-wp-fonts-library.php:
```diff
@@ -0,0 +1,618 @@
+            $new_library_fonts = $this->merge_fonts( $font_families, $new_fonts );
+
+            // Sanitizes the fonts library using WP_Theme_JSON
+            $sanitized_font_families = $this->sanitize_font_families( $new_library_fonts );
```
`$sanitized_font_families` is not used and can be removed.
```suggestion
            $this->sanitize_font_families( $new_library_fonts );
```

anton-vlasenko commented on lib/experimental/fonts-library/class-wp-fonts-library.php:
```diff
@@ -0,0 +1,618 @@
+            return true;
+        } else {
+            return $this->delete_asset( $font_face['src'] );
+        }
```
Let's [avoid nesting and return early](https://github.com/piotrplenik/clean-code-php#avoid-nesting-too-deeply-and-return-early-part-1):
```suggestion
        if ( ! is_array( $font_face['src'] ) ) {
            return $this->delete_asset( $font_face['src'] );
        }
        
        foreach ( $font_face['src'] as $src ) {
            $was_assets_removed = $this->delete_asset($src);
            if ( ! $was_assets_removed ) {
                // Bail if any of the assets could not be removed
                return false;
            }
        }
        return true;
        
```

matiasbenedetto replied:
changed here: https://github.com/WordPress/gutenberg/pull/50927/commits/645626f1f8b9206c268a4dbc7f3b05ea25e17ac6

anton-vlasenko commented on lib/experimental/fonts-library/class-wp-fonts-library.php:
```diff
@@ -0,0 +1,618 @@
+        $filename =  $this->get_filename_from_font_face( $font_face, $file['name'] );
+        $filepath= path_join( $this->wp_fonts_dir, $filename );
+
+        // Remove the uploaded font asset reference from the font face definition because is not longer needed
```
Nitpick:
```suggestion
        // Remove the uploaded font asset reference from the font face definition because it is no longer needed.
```

anton-vlasenko commented on lib/experimental/fonts-library/class-wp-fonts-library.php:
```diff
@@ -0,0 +1,618 @@
+        $new_fonts = $this->download_or_move_fonts( $fonts_to_install, $files );
+
+        if ( ! empty ( $new_fonts )  ){
+            // Updates the fonts library with the new fonts succesfully downloaded
```
Nitpick:
```suggestion
            // Updates the fonts library with the new successfully downloaded fonts.
```

anton-vlasenko commented on lib/experimental/fonts-library/class-wp-fonts-library.php:
```diff
@@ -0,0 +1,618 @@
+
+        // Moves the file to the fonts directory or return false
+        $renamed_file = rename( $temp_file, $file_path );
+        if ( ! $renamed_file ) {
+            @unlink( $temp_file );
+            return false;
+        }
+
+        // Cleans the temp file
+        @unlink( $temp_file );
```
It seems that `$temp_file` is being deleted regardless of the value of `$renamed_file`, so I would refactor that code.
```suggestion
        // Cleans the temp file
        @unlink( $temp_file );
        
        if ( ! $renamed_file ) {            
            return false;
        }
```

matiasbenedetto replied:
Implemented here: https://github.com/WordPress/gutenberg/pull/50927/commits/645626f1f8b9206c268a4dbc7f3b05ea25e17ac6

anton-vlasenko commented on lib/experimental/fonts-library/class-wp-fonts-library.php:
```diff
@@ -0,0 +1,618 @@
+                        $new_font_face = $this->move_font_face_asset( $font_face, $files[ $font_face[ 'file' ] ] );
+                    }
+                    
+                    // If the font face assets were succesfully downloaded, we add the font face to the new font
```
Nitpick:
```suggestion
                    // If the font face assets were successfully downloaded, we add the font face to the new font.
```

anton-vlasenko commented on lib/experimental/fonts-library/class-wp-fonts-library.php:
```diff
@@ -0,0 +1,618 @@
+                $new_font['fontFace'] = $new_font_faces;
+            }
+
+            // If the font face assets were succesfully downloaded, we add the font to the new fonts array
```
Nitpick:
```suggestion
            // If the font face assets were successfully downloaded, we add the font to the new fonts array.
```

anton-vlasenko commented on lib/experimental/fonts-library/class-wp-fonts-library.php:
```diff
@@ -0,0 +1,618 @@
+        $file_was_moved = move_uploaded_file( $file['tmp_name'], $filepath );
+        
+        if ( $file_was_moved ) {
+            // If the file was succesfully moved, we update the font face definition to reference the new file location
```
Nitpick:
```suggestion
            // If the file was successfully moved, we update the font face definition to reference the new file location
```

anton-vlasenko commented on lib/experimental/fonts-library/class-wp-fonts-library.php:
```diff
@@ -0,0 +1,618 @@
+        $weight = sanitize_title( $font_face['fontWeight'] );
+        $filename = "{$family}_{$style}_{$weight}";
+        if ($i > 0) {
+            $filename = "{$filename}_{$i}";
```
This could be replaced with
```suggestion
            $filename .= "_{$i}";
```

anton-vlasenko commented on lib/experimental/fonts-library/class-wp-fonts-library.php:
```diff
@@ -0,0 +1,618 @@
+        $new_font_face = $font_face;
+        if ( is_array( $font_face['src'] ) ) {
+            $new_font_face['src'] = array();
+            $i;
+            foreach ( $font_face['src'] as $src ) {
+                !$i ? $i = 0 : $i++;
+                $filename = $this->get_filename_from_font_face( $font_face, $src, $i );
```
I don't understand the need to check the value of `$i` with the ternary operator.
IMO this could be replaced with: 
```suggestion
            $i = 0;
            foreach ( $font_face['src'] as $src ) {              
                $filename = $this->get_filename_from_font_face( $font_face, $src, ++$i );
```

anton-vlasenko commented on lib/experimental/fonts-library/class-wp-fonts-library.php:
```diff
@@ -0,0 +1,618 @@
+     * @param array $font_face The font face array containing the 'src' attribute with the source URL(s) of the assets.
+     * @return array The modified font face array with the new source URL(s) to the downloaded assets.
+     */
+    function download_font_face_assets ( $font_face ) {
```
This code appears to be duplicated:
```php
$filename = $this->get_filename_from_font_face( $font_face, $font_face['src'] );
$new_src = $this->download_asset( $font_face['src'], $filename );
```

I'd move it to a separate class method to avoid code duplication.

anton-vlasenko commented on lib/experimental/fonts-library/class-wp-fonts-library.php:
```diff
@@ -0,0 +1,618 @@
+ * @since x.x.x
+ * 
+ */
+class WP_Fonts_Library extends WP_REST_Controller {
```
Classes that extend the WP_REST_Controller class typically have the suffix "_Controller" in their name as well.
```suggestion
class WP_Fonts_Library_Controller extends WP_REST_Controller {
```

anton-vlasenko reviewed (commented):
I've done the initial code review, @matiasbenedetto.
Please find my notes below.
Additionally, it would be great to:
- add PHPUnit tests for the `WP_Fonts_Library` controller;
- format PHPDOc blocks according to Core standards (i.e. add indentation for parameter types and return values);

These changes will ensure that the code is more Core-friendly.

matiasbenedetto committed:
Add _Controller to PHP class name

Co-authored-by: Anton Vlasenko <43744263+anton-vlasenko@users.noreply.github.com>

matiasbenedetto committed:
comment spell fix

Co-authored-by: Anton Vlasenko <43744263+anton-vlasenko@users.noreply.github.com>

matiasbenedetto committed:
comment spell fix

Co-authored-by: Anton Vlasenko <43744263+anton-vlasenko@users.noreply.github.com>

matiasbenedetto committed:
minor syntax change

Co-authored-by: Anton Vlasenko <43744263+anton-vlasenko@users.noreply.github.com>

matiasbenedetto committed:
comment spelling fix

Co-authored-by: Anton Vlasenko <43744263+anton-vlasenko@users.noreply.github.com>

matiasbenedetto committed:
comment spelling

Co-authored-by: Anton Vlasenko <43744263+anton-vlasenko@users.noreply.github.com>

matiasbenedetto committed:
comment spelling

Co-authored-by: Anton Vlasenko <43744263+anton-vlasenko@users.noreply.github.com>

matiasbenedetto committed:
updating class name

matiasbenedetto committed:
Refactoring the use of global settings state

matiasbenedetto committed:
TEMPORAL FIX: Disabling add_missing_fonts_to_theme_json() untils this is fixed in the Fonts API

matiasbenedetto committed:
public php class method

Co-authored-by: Anton Vlasenko <43744263+anton-vlasenko@users.noreply.github.com>

matiasbenedetto commented:
Linking an issue we would need to fix to make this the editor preview without reloading the page when adding new fonts: https://github.com/WordPress/gutenberg/issues/51764

matiasbenedetto committed:
Merge branch 'trunk' into try/font-library

matiasbenedetto committed:
fix after merging latest changes from trunk

matiasbenedetto committed:
Declaring class properties

matiasbenedetto committed:
minor refactor of delete_font_face_assets and delete_asset functions

matiasbenedetto committed:
removing line not needed

matiasbenedetto committed:
Removing not needed variable definition

Co-authored-by: Anton Vlasenko <43744263+anton-vlasenko@users.noreply.github.com>

matiasbenedetto committed:
improving syntax

Co-authored-by: Anton Vlasenko <43744263+anton-vlasenko@users.noreply.github.com>

matiasbenedetto committed:
avoid redundant code line

Co-authored-by: Anton Vlasenko <43744263+anton-vlasenko@users.noreply.github.com>

matiasbenedetto committed:
avoiding duplicated code

hellofromtonya commented:
Heads Up: The new [Font Face](https://github.com/WordPress/gutenberg/pull/51770) was merged into `trunk` today.

Rebasing this PR on top of `trunk` will bring in the Font Face and disable the Fonts API. Then further testing can happen within the scope of the Fonts Library.

What is Font Face? See https://github.com/WordPress/gutenberg/issues/45271#issuecomment-1629438025

matiasbenedetto committed:
Refactoring to use one post per font family in the library

matiasbenedetto committed:
Refactoring delete font family

matiasbenedetto committed:
avoid adding number to filename when there is only one asset

matiasbenedetto committed:
Refactor install fonts endpoint to understand only formdata

matiasbenedetto committed:
Renaming PHP file

matiasbenedetto committed:
Spliting backend code in 2 clasess and refactoring

matiasbenedetto committed:
hide preview tools on local fonts upload

matiasbenedetto committed:
Merge branch 'trunk' into try/font-library

matiasbenedetto committed:
Avoid tracking post id in font family class

jffng committed:
Format php.

jffng committed:
Update comment.

jffng committed:
Fix array key issue with merge.

matiasbenedetto commented:
We are splitting the font library in several PRs to merge it progressively:
The backend code for the font library is available for review in this PR https://github.com/WordPress/gutenberg/pull/52704

matiasbenedetto committed:
rename custom post type to wp_font_family

paaljoachim commented:
Hey @matiasbenedetto 

Really nice work!

Just a quick comment after seeing the first video.

An example.
I install a lot of fonts and variations of them. After trying them out I find 2 fonts I want to keep and the rest I just want to delete. From the first video it looks like we can deselect but not delete a font. There can now be a huge list of fonts that are deselected and still seen in that specific screen.

It would be helpful to select - deselect as well as delete a font when one knows it will not be used.

matiasbenedetto commented:
> It would be helpful to select - deselect as well as delete a font when one knows it will not be used.

Hi @paaljoachim, yes! I think it does not appear on the screencast, but currently, there is a "uninstall" button available for the font families with no font faces activated. When clicked, It will remove the font definition and the font assets.

paaljoachim commented:
Great! Thank you @matiasbenedetto Matias!

azaozz reviewed (changes requested):
A question: Is there a specific reason for the Font Library to use a third-party proprietary CDN for fonts sources? Asking because use of Google Fonts was one of the reason the initial "Webfonts API" was not committed to WordPress two years ago. Then there was a GDPR related court case about usage of Google Fonts where, as far as I remember, that usage was deemed inappropriate/breaking the GDPR. There seem to be other sources for the same fonts.

Seems use of the Google Fonts CDN in WP should be discouraged, and core should not force any website to it. What is the chance to remove use of the huge (1.3MB) [list of available Google fonts](https://github.com/WordPress/gutenberg/blob/6b76e479fb7eb99ee74cae6e676e62cddeb153fc/lib/experimental/fonts-library/google-fonts.json) and replace it with a list to the same fonts hosted elsewhere? As far as I see there are several places that host the same open source fonts, even as NPM packages (would that be a preferable format?).

ResponsiveFlow commented:
@azaozz I have the same doubts regarding Google Fonts and GDPR compliance.
Why not creating something like Openverse but for fonts?
Or why not integrating open source fonts in Openverse?

ResponsiveFlow commented:
> there is a "uninstall" button available for the font families with no font faces activated

@matiasbenedetto I think that, for font families with active font faces, the button should be displayed anyway, but grayed out and with a hint saying that it is is not possible to delete that font because it is currently active somewhere(?).

matiasbenedetto commented:
> @matiasbenedetto I think that, for font families with active font faces, the button should be displayed anyway, but grayed out and with a hint saying that it is is not possible to delete that font because it is currently active somewhere(?).

Hi @tomxygen, Yes! Good suggestion.
This was implemented in the [latest frontend PR](https://github.com/WordPress/gutenberg/pull/53884) around [Font Library](https://github.com/WordPress/gutenberg/issues/54169): 

I'm closing this PR because it was just a prototype. 
The project is being tracked here: https://github.com/WordPress/gutenberg/issues/54169
